crates/sui-types/src/attestation.rs

```diff
@@ -0,0 +1,432 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::time::SystemTime;
+
+use crate::error::{SuiError, SuiResult};
+use std::collections::BTreeMap;
+
+use boring_signal::bn::BigNum;
+use boring_signal::ecdsa::EcdsaSig;
+use boring_signal::stack;
+use boring_signal::x509::store::X509StoreBuilder;
+use boring_signal::x509::{X509StoreContext, X509};
+use ciborium::value::{Integer, Value};
+use prost::DecodeError;
+use sha2::{Digest, Sha384};
+use subtle::ConstantTimeEq;
+#[cfg(test)]
+#[path = "unit_tests/attestation_tests.rs"]
+mod attestation_tests;
+
+const ROOT_CERTIFICATE_PEM: &[u8] = include_bytes!("./nitro_root_certificate.pem");
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum NitroError {
+    /// Invalid CBOR
+    InvalidCbor,
+    /// Invalid COSE_Sign1
+    InvalidCoseSign1,
+    /// Invalid signature
+    InvalidSignature,
+    /// Invalid attestation document
+    InvalidAttestationDoc,
+    /// Invalid certificate: {0}
+    InvalidCertificate(String),
+    /// Invalid PCRs
+    InvalidPcrs,
+    /// Invalid Public Key
+    InvalidPublicKey,
+    /// User data field is absent from the attestation document
+    UserDataMissing,
+    /// Invalid User Data
+    InvalidUserData,
+}
+
+impl From<ciborium::de::Error<std::io::Error>> for NitroError {
+    fn from(_err: ciborium::de::Error<std::io::Error>) -> NitroError {
+        NitroError::InvalidCbor
+    }
+}
+
+impl From<boring_signal::error::ErrorStack> for NitroError {
+    fn from(err: boring_signal::error::ErrorStack) -> NitroError {
+        NitroError::InvalidCertificate(err.to_string())
+    }
+}
+
+impl From<DecodeError> for NitroError {
+    fn from(_err: DecodeError) -> Self {
+        NitroError::InvalidUserData
+    }
+}
+
+/// <https://docs.aws.amazon.com/enclaves/latest/user/verify-root.html>
+pub fn attestation_verify_inner(
+    document_data: &[u8],
+    pk: &[u8],
+    pcr0: &[u8],
+    pcr1: &[u8],
+    pcr2: &[u8],
+) -> SuiResult<()> {
+    // Following the steps here: https://docs.aws.amazon.com/enclaves/latest/user/verify-root.html
+    // Step 1. Decode the CBOR object and map it to a COSE_Sign1 structure
+    let cose_sign1 = CoseSign1::from_bytes(document_data)
+        .map_err(|_| SuiError::AttestationFailedToVerify("cannot parse cose sign1".to_string()))?;
+    let now = SystemTime::now();
```

Comment:
nitro enclave attestation expires in 3 hours. is it safe to use system time now from validators?

Reply:
local time is not consistent, we can read Clock on the Move level and pass its value to the native call